### PR TITLE
Feat/string bytes

### PR DESCRIPTION
**Pull request by ccamel**
## Scope

This PR introduces implements the standard predicate `string_bytes/3` as specified by #497 and functionally aligned with the [SWI-Prolog implementation ](https://www.swi-prolog.org/pldoc/doc_for?object=string_bytes/3)(except for errors).

## Implementation details

The PR comes with many changes, my apologies for that. But the implementation required a refactoring of how prolog terms are processed during conversion to and from byte list representations. This involved introducing support for various encodings (e.g., UTF-8).

Additionally, a first step towards standardization of error messages has been undertaken. This primarily involves transitioning from text-described errors to errors represented by Prolog terms. These terms are highly compatible with the conventions outlined in [SWI-Prolog's documentation](https://www.swi-prolog.org/pldoc/man?section=error). This foundation for the approach exists in [Itchiban Prolog](https://github.com/ichiban/prolog).

Finally, this PR lays the groundwork for implementing additional conversion predicates, particularly those related to `base64` and `base58` encodings.

## Summary by CodeRabbit

- **New Features**
  - Introduced new search and match capabilities with Prolog integration.
  - Enhanced JSON handling and conversion within Prolog terms.
  - Added support for advanced string and byte conversions.
  - Improved hexadecimal encoding and decoding functions.
  
- **Improvements**
  - Updated error messaging for clarity and specificity in test functions.
  - Refined balance fetching logic with Prolog package utilization.
  
- **Bug Fixes**
  - Fixed issues with incorrect package references, ensuring consistency across the codebase.
  
- **Refactor**
  - Replaced deprecated utility functions with streamlined Prolog package methods.
  - Removed redundant code for a cleaner and more efficient codebase.
  
- **Tests**
  - Expanded test coverage for new and existing functionalities, including encoding and JSON term extraction.
  
- **Documentation**
  - Updated references to reflect the shift from utility to Prolog package use across various modules.

### CODE REVIEW

**Review by amimart — APPROVED** [review]
That was huge! Seems clear to me, hopefully tests are here to help me gain confidence 😄

**Comment by ccamel** [comment]
Thanks so much for your review @amimart. Indeed, it gets tricky and tests are definitely essential.

File: x/logic/prolog/term_test.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+							Convey("and should be as expected", func() {
+								So(err.Error(), ShouldEqual, tc.wantError.Error())
+							})
+						})
+					}
+				})
+			})
+		}
+	})
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/prolog/term_test.go:60]: Error will be standardized by #499.

File: x/logic/prolog/term_test.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/prolog/term_test.go:193]: Error will be standardized by #499.

File: x/logic/predicate/string_test.go [hunk 3]
```diff
@@ -197,3 +197,163 @@ func TestReadString(t *testing.T) {
...
+									})
+								})
+							})
+						})
+					})
+				})
+			})
+		}
+	})
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/string_test.go:359]: Error will be standardized by #499.

File: x/logic/predicate/crypto.go [hunk 4]
```diff
@@ -69,7 +69,7 @@ func CryptoDataHash(
 				algorithmAtom.String(),
 				util.HashAlgNames()))
 		}
-		decodedData, err := TermToBytes(data, options, AtomUtf8, env)
+		decodedData, err := termToBytes(data, options, prolog.AtomUtf8, env)
 		if err != nil {
 			return engine.Error(fmt.Errorf("%s: failed to decode data: %w", functor, err))
 		}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/crypto.go:75]: Error will be standardized by #499.

File: x/logic/predicate/crypto.go [hunk 5]
```diff
@@ -239,17 +184,17 @@
...
+		decodedData, err := termToBytes(data, options, prolog.AtomHex, env)
 		if err != nil {
 			return engine.Error(fmt.Errorf("%s: failed to decode data: %w", functor, err))
 		}
 
-		decodedSignature, err := TermToBytes(sig, AtomEncoding.Apply(AtomOctet), AtomHex, env)
+		decodedSignature, err := termToBytes(sig, prolog.AtomEncoding.Apply(prolog.AtomOctet), prolog.AtomHex, env)
 		if err != nil {
 			return engine.Error(fmt.Errorf("%s: failed to decode signature: %w", functor, err))
 		}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/crypto.go:200]: Error will be standardized by #499.

File: x/logic/predicate/crypto.go [hunk 6]
```diff
@@ -267,19 +212,22 @@
...
 
-	return atom, nil
+	switch encodingAtom {
+	case prolog.AtomHex:
+		return prolog.TermHexToBytes(term, env)
+	case prolog.AtomOctet, prolog.AtomUtf8:
+		return prolog.StringTermToBytes(term, prolog.AtomEmpty, env)
+	default:
+		return nil, fmt.Errorf("invalid encoding: %s. Possible values: hex, octet", encodingAtom.String())
+	}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/predicate/crypto.go:232]: Error will be standardized by #499.

File: x/logic/prolog/term.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+		terms = append(terms, engine.Integer(arg))
+	}
+	return engine.List(terms...)
+}
+
+// StringToTerm converts a string to a term.
+// TODO: this function should be removed.
+func StringToTerm(s string) engine.Term {
+	return engine.NewAtom(s)
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/prolog/term.go:28]: Yes, this function will be removed after a slight clean-up of the code. I didn't want to make too many changes to this PR.

**Pull request merged**